sql/sql_statistics.cc

```diff
@@ -1651,83 +1664,81 @@ std::vector<std::string> bucket_bounds = {};
 
     test_parse_histogram_from_json();
   }
+};
 
-  static std::vector<std::string> parse_histogram_from_json(const char *json)
-  {
-    std::vector<std::string> hist_buckets= {};
-    enum json_types vt = json_get_array_items(json, json + strlen(json), hist_buckets);
-    printf("%d", vt);
-    printf("%zu", hist_buckets.size());
+void test_parse_histogram_from_json()
+{
+  std::vector<std::string> bucket = {};
+  std::string json;
+  std::string tests[7] = {
+      R"(["aabbb", "ccccdd", "eeefff"])",
+      R"(["aabbb", "ccc{}dd", "eeefff"])",
+      R"(["aabbb", {"a": "b"}, "eeefff"])",
+      R"({})",
+      R"([1,2,3, null])",
+      R"([null])",
+      R"([])"
+  };
 
-    return hist_buckets;
+  for(const auto& test : tests) {
+    json = test;
+    bucket = parse_histogram_from_json(json.c_str());
   }
+}
 
-  static void test_parse_histogram_from_json()
-  {
-    std::vector<std::string> bucket = {};
-    std::string json;
-    std::string tests[7] = {
-        R"(["aabbb", "ccccdd", "eeefff"])",
-        R"(["aabbb", "ccc{}dd", "eeefff"])",
-        R"(["aabbb", {"a": "b"}, "eeefff"])",
-        R"({})",
-        R"([1,2,3, null])",
-        R"([null])",
-        R"([])"
-    };
-
-    for(const auto& test : tests) {
-      json = test;
-      bucket = parse_histogram_from_json(json.c_str());
-      printf("%zu", bucket.size());
-    }
-  }
+std::vector<std::string> parse_histogram_from_json(const char *json)
+{
+  std::vector<std::string> hist_buckets= {};
+  int vt;
+  bool result = json_get_array_items(json, json + strlen(json), &vt, hist_buckets);
+  fprintf(stderr,"==============\n");
+  fprintf(stderr,"histogram: %s\n", json);
+  fprintf(stderr, "json_get_array_items() returned %s\n", result ? "true" : "false");
+  fprintf(stderr, "value type after json_get_array_items() is %d\n", vt);
+  fprintf(stderr, " JSV_BAD_JSON=%d, JSON_VALUE_ARRAY=%d\n", (int)JSV_BAD_JSON, (int)JSON_VALUE_ARRAY);
+  fprintf(stderr, "hist_buckets.size()=%zu\n", hist_buckets.size());
+
+  return hist_buckets;
+}
 
-  /*
-   * json_get_array_items expects a JSON array as argument,
-   * and pushes the elements of the array into the `container` vector.
-   * It only works if all the elements in the original JSON array
-   * are scalar values (i.e., strings, numbers, true or false), and returns JSV_BAD_JSON if:
-   * the original JSON is not an array OR the JSON array contains non-scalar elements.
-   */
-  static json_types json_get_array_items(const char *json, const char *json_end, std::vector<std::string> &container) {
-    json_engine_t je;
-    enum json_types value_type;
-    int vl;
-    const char *v;
+bool json_get_array_items(const char *json, const char *json_end, int *value_type, std::vector<std::string> &container) {
+  json_engine_t je;
```

Comment:
I don't understand, what is the use of `value_type` for the caller?

Reply:
It's a way of telling the caller why parsing failed. That way, they know how the JSON structure was interpreted.